File: sdks/python/dcp_ai/providers/ml_dsa_65_provider.py
"""
ML-DSA-65 signature provider using pqcrypto (PQClean bindings).

FIPS 204, NIST Level 3. Primary post-quantum signature algorithm.
Public key: 1952 B, Signature: 3309 B.
"""

from __future__ import annotations

import base64

from pqcrypto.sign.dilithium3 import (
    SIGNATURE_SIZE,
    generate_keypair,
    sign as pqc_sign,
    verify as pqc_verify,
)

from dcp_ai.v2.crypto_provider import CryptoProvider, derive_kid


class MlDsa65Provider(CryptoProvider):
    alg = "ml-dsa-65"
    key_size = 1952
    sig_size = 3309
    is_constant_time = True

    async def generate_keypair(self) -> dict[str, str]:
        pk, sk = generate_keypair()
        pk_bytes = bytes(pk)
        kid = derive_kid(self.alg, pk_bytes)
        return {
            "kid": kid,
            "public_key_b64": base64.b64encode(pk_bytes).decode("ascii"),
            "secret_key_b64": base64.b64encode(bytes(sk)).decode("ascii"),
        }

    async def sign(self, message: bytes, secret_key_b64: str) -> bytes:
        sk_bytes = base64.b64decode(secret_key_b64)
        signed_msg = pqc_sign(sk_bytes, message)
        return bytes(signed_msg[:SIGNATURE_SIZE])

    async def verify(self, message: bytes, signature: bytes, public_key_b64: str) -> bool:
        pk_bytes = base64.b64decode(public_key_b64)
        signed_msg = bytes(signature) + message
        try:
            pqc_verify(pk_bytes, signed_msg)
            return True
        except Exception:
            return False

File: sdks/go/dcp/providers/ml_dsa_65_provider.go
package providers

import (
	"encoding/base64"
	"fmt"

	"github.com/cloudflare/circl/sign/mldsa/mldsa65"

	v2 "github.com/dcp-ai-protocol/dcp-ai/sdks/go/v2/dcp/v2"
)

// MlDsa65Provider implements v2.CryptoProvider using ML-DSA-65 (FIPS 204, Level 3).
type MlDsa65Provider struct{}

func (p *MlDsa65Provider) Alg() string         { return "ml-dsa-65" }
func (p *MlDsa65Provider) KeySize() int         { return 1952 }
func (p *MlDsa65Provider) SigSize() int         { return 3309 }
func (p *MlDsa65Provider) IsConstantTime() bool { return true }

func (p *MlDsa65Provider) GenerateKeypair() (*v2.GeneratedKeypair, error) {
	pk, sk, err := mldsa65.GenerateKey(nil)
	if err != nil {
		return nil, fmt.Errorf("ml-dsa-65 keygen: %w", err)
	}
	pkBytes, err := pk.MarshalBinary()
	if err != nil {
		return nil, fmt.Errorf("ml-dsa-65 marshal pk: %w", err)
	}
	skBytes, err := sk.MarshalBinary()
	if err != nil {
		return nil, fmt.Errorf("ml-dsa-65 marshal sk: %w", err)
	}
	kid := v2.DeriveKid("ml-dsa-65", pkBytes)
	return &v2.GeneratedKeypair{
		Kid:          kid,
		PublicKeyB64: base64.StdEncoding.EncodeToString(pkBytes),
		SecretKeyB64: base64.StdEncoding.EncodeToString(skBytes),
	}, nil
}

func (p *MlDsa65Provider) Sign(message []byte, secretKeyB64 string) ([]byte, error) {
	skBytes, err := base64.StdEncoding.DecodeString(secretKeyB64)
	if err != nil {
		return nil, fmt.Errorf("decode secret key: %w", err)
	}
	var sk mldsa65.PrivateKey
	if err := sk.UnmarshalBinary(skBytes); err != nil {
		return nil, fmt.Errorf("unmarshal ml-dsa-65 sk: %w", err)
	}
	sig := make([]byte, mldsa65.SignatureSize)
	if err := mldsa65.SignTo(&sk, message, nil, false, sig); err != nil {
		return nil, fmt.Errorf("ml-dsa-65 sign: %w", err)
	}
	return sig, nil
}

func (p *MlDsa65Provider) Verify(message []byte, signature []byte, publicKeyB64 string) (bool, error) {
	pkBytes, err := base64.StdEncoding.DecodeString(publicKeyB64)
	if err != nil {
		return false, fmt.Errorf("decode public key: %w", err)
	}
	var pk mldsa65.PublicKey
	if err := pk.UnmarshalBinary(pkBytes); err != nil {
		return false, fmt.Errorf("unmarshal ml-dsa-65 pk: %w", err)
	}
	return mldsa65.Verify(&pk, message, nil, signature), nil
}

var _ v2.CryptoProvider = (*MlDsa65Provider)(nil)
